File: python/camera_ui_sdk/camera/events.py
from __future__ import annotations

from typing import Literal, NotRequired, TypedDict

from ..sensor.detection import BoundingBox
from .enums import DetectionEventType, EventTriggerType


class EventDetection(TypedDict):
    """Aggregated object detection within a segment."""

    label: str
    """Detection label (e.g. "person", "car")."""
    score: float
    """Best confidence score."""
    maxCount: int
    """Maximum simultaneous count in a single frame."""
    box: NotRequired[BoundingBox]
    """Bounding box of the highest-confidence detection (normalized 0-1)."""
    thumbnail: NotRequired[bytes]
    """Best-selected JPEG thumbnail crop. Only present on 'end' events."""
    trackId: NotRequired[int]
    """Object tracker ID (links this detection across frames)."""
    moving: NotRequired[bool]
    """Whether the object was moving (True) or stationary (False)."""


class EventAttribute(TypedDict):
    """Unified attribute within a segment (face identity, license plate, classifier result)."""

    type: str
    """Attribute type ('face', 'license_plate', or classifier-specific like 'bird')."""
    label: str
    """Identity name, plate text, or classification label."""
    confidence: NotRequired[float]
    """Detection confidence (0-1)."""
    thumbnail: NotRequired[bytes]
    """Best-selected JPEG thumbnail crop. Only present on 'end' events."""
    embedding: NotRequired[list[float]]
    """Face embedding vector for unknown face persistence."""
    embeddingModel: NotRequired[str]
    """Embedding model identifier."""
    clipEmbedding: NotRequired[list[float]]
    """CLIP embedding vector for semantic search."""
    clipEmbeddingModel: NotRequired[str]
    """CLIP embedding model identifier."""
    parentTrackId: NotRequired[int]
    """Parent object's tracker ID (links this attribute to its parent detection)."""


class EventTrigger(TypedDict):
    """Event trigger (motion, audio, sensor, or line-crossing)."""

    type: EventTriggerType
    """Trigger type."""
    label: NotRequired[str]
    """Audio label (e.g. "doorbell", "glass_break")."""
    score: NotRequired[float]
    """Best confidence score."""
    firstSeen: int
    """First detection time (Unix ms)."""
    lastSeen: int
    """Last detection time (Unix ms)."""
    lineName: NotRequired[str]
    """Name of the crossed line (only for line-crossing triggers)."""
    crossingDirection: NotRequired[str]
    """Crossing direction (only for line-crossing triggers)."""
    trackId: NotRequired[int]
    """Track ID of the object that crossed (only for line-crossing triggers)."""


class EventDescription(TypedDict):
    """AI-generated event description."""

    title: str
    """Brief title of what occurred."""
    description: str
    """Chronological narrative of the scene."""
    summary: str
    """Two-sentence notification-friendly summary."""
    threatLevel: int
    """Threat level: 0 = normal, 1 = suspicious, 2 = threat."""


class EventSegment(TypedDict):
    """A contiguous object detection phase within an event."""

    firstSeen: int
    """Segment start time (Unix ms)."""
    lastSeen: int
    """Segment end time (Unix ms)."""
    thumbnail: NotRequired[bytes]
    """Best-selected JPEG scene thumbnail for this segment. Only present on 'end' events."""
    detections: list[EventDetection]
    """Object detections in this segment."""
    attributes: list[EventAttribute]
    """Unified attributes (faces, plates, classifications)."""
    zones: NotRequired[list[str]]
    """Names of detection zones any detection in this segment overlapped (deduplicated)."""
    description: NotRequired[EventDescription]
    """AI-generated description for this segment."""


class DetectionEvent(TypedDict):
    """Aggregated detection event with lifecycle (start -> update -> end).
    Groups individual sensor detections into structured events."""

    id: str
    """Unique event ID."""
    cameraId: str
    """Camera that produced this event."""
    state: Literal["active", "ended"]
    """Event lifecycle state."""
    startTime: int
    """Event start time (Unix ms)."""
    endTime: NotRequired[int]
    """Event end time (Unix ms, only when ended)."""
    lastUpdate: int
    """Last activity timestamp (Unix ms)."""
    types: list[str]
    """Detection types present in this event (for filtering)."""
    triggers: list[EventTrigger]
    """Event triggers (motion/audio)."""
    segments: list[EventSegment]
    """Detection segments (object detection phases).
    For segment-* messages: contains only the current segment.
    For start/end messages: empty."""
    segmentIndex: NotRequired[int]
    """Index of the segment in segments[0] for segment-* messages."""
    expectedEndTime: NotRequired[int]
    """Expected event end time (Unix ms) — the latest dwell expiry across all
    currently-active triggers. Monotonically non-decreasing during the event
    lifetime. Updated on each `update` / `segment-*` message."""
    thumbnail: NotRequired[bytes]
    """Full-frame downscaled JPEG captured at event start. Inline only on the first
    message that delivers it (start or the first update); the NVR plugin persists it
    and clients fetch it on demand via getEventThumbnails."""


class DetectionEventPayload(TypedDict):
    """Emitted for detection events (start/update/end)."""

    type: DetectionEventType
    """Event lifecycle type."""
    event: DetectionEvent
    """Full event data."""

File: node/src/camera/events.ts
import type { DetectionEventState, EventTriggerType } from './enums.js';

/** All event trigger types as a runtime-accessible array. */
export const EVENT_TRIGGER_TYPES = [
  'motion',
  'audio',
  'contact',
  'doorbell',
  'switch',
  'light',
  'siren',
  'security_system',
  'line-crossing',
  'occupancy',
  'smoke',
  'leak',
  'gas',
  'carbonMonoxide',
  'heat',
  'cold',
  'vibration',
  'tamper',
  'problem',
  'power',
  'lock',
  'garage',
] as const satisfies readonly EventTriggerType[];

/** Event trigger (motion, audio, sensor, or line-crossing). */
export interface EventTrigger {
  /** Trigger type. */
  type: EventTriggerType;
  /** Audio label (e.g. "doorbell", "glass_break"). */
  label?: string;
  /** Best confidence score. */
  score?: number;
  /** First detection time (Unix ms). */
  firstSeen: number;
  /** Last detection time (Unix ms). */
  lastSeen: number;
  /** Name of the crossed line (only for line-crossing triggers). */
  lineName?: string;
  /** Crossing direction (only for line-crossing triggers). */
  crossingDirection?: 'a-to-b' | 'b-to-a';
  /** Track ID of the object that crossed (only for line-crossing triggers). */
  trackId?: number;
}

/** Aggregated object detection within a segment. */
export interface EventDetection {
  /** Detection label (e.g. "person", "car"). */
  label: string;
  /** Best confidence score. */
  score: number;
  /** Maximum simultaneous count in a single frame. */
  maxCount: number;
  /** Whether the object was moving (true) or stationary (false). */
  moving?: boolean;
  /** Where the tracked object entered and left the frame within this segment. */
  path?: DetectionPath;
  /** Names of detection/alert zones any object of this label overlapped during the segment. */
  zones?: string[];
}

/** Normalized box centers (0-1) of a tracked object's first and last sighting in a segment. */
export interface DetectionPath {
  enterX: number;
  enterY: number;
  exitX: number;
  exitY: number;
}

/** Unified attribute within a segment (face identity, license plate, classifier result). */
export interface EventAttribute {
  /** Attribute type ('face', 'license_plate', or classifier-specific like 'bird'). */
  type: string;
  /** Identity name, plate text, or classification label. */
  label: string;
  /** Detection confidence (0-1). */
  confidence?: number;
}

/** A contiguous object detection phase within an event. */
export interface EventSegment {
  /** Segment start time (Unix ms). */
  firstSeen: number;
  /** Segment end time (Unix ms). */
  lastSeen: number;
  /** Object detections in this segment. */
  detections: EventDetection[];
  /** Unified attributes (faces, plates, classifications). */
  attributes: EventAttribute[];
  /** Names of detection zones any detection in this segment overlapped (deduplicated). */
  zones?: string[];
}

/**
 * Aggregated detection event with lifecycle (start -> update -> end).
 * Groups individual sensor detections into structured events.
 */
export interface DetectionEvent {
  /** Unique event ID. */
  id: string;
  /** Camera that produced this event. */
  cameraId: string;
  /** Event lifecycle state. */
  state: DetectionEventState;
  /** Event start time (Unix ms). */
  startTime: number;
  /** Event end time (Unix ms, only when ended). */
  endTime?: number;
  /** Last activity timestamp (Unix ms). */
  lastUpdate: number;
  /** Detection types present in this event (for filtering). */
  types: string[];
  /** Event triggers (motion/audio). */
  triggers: EventTrigger[];
  /**
   * Detection segments (object detection phases). For segment-* messages this
   * contains only the current segment, for start/end messages it is empty.
   */
  segments: EventSegment[];
  /** Index of the segment in segments[0] for segment-* messages. */
  segmentIndex?: number;
  /**
   * Expected event end time (Unix ms): the latest dwell expiry across all
   * currently-active triggers. Monotonically non-decreasing during the event
   * lifetime. Updated on each `update` / `segment-*` message.
   */
  expectedEndTime?: number;
}
